### PR TITLE
BUG: np.histogramdd loses precision on its inputs, leading to incorrect results

### PR DESCRIPTION
**Pull request by eric-wieser**
Previously:

* The last bin edge was treated as "fuzzy", extending 1e-6 times the smallest bin width beyond the end of the array (gh-10864)
* Integer bin edges were cast to float, which for large integers would cause non-equal edges to become equal

Note that this is different to gh-7845, which is about precision on the _outputs_.

### CODE REVIEW

File: numpy/lib/tests/test_histograms.py [hunk 1]
```diff
@@ -660,3 +658,40 @@ def test_finite_range(self):
                       range=[[0.0, 1.0], [0.25, 0.75], [0.25, np.inf]])
         assert_raises(ValueError, histogramdd, vals,
                       range=[[0.0, 1.0], [np.nan, 0.75], [0.25, 0.5]])
+
+    def test_equal_edges(self):
```
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: On master, this test crashes with
```
ValueError: `bins[0]` must be strictly increasing, when an array
```

File: numpy/lib/tests/test_histograms.py [hunk 2]
```diff
@@ -660,3 +658,40 @@ def test_finite_range(self):
...
+    def test_edge_dtype(self):
+        """ Test that if an edge array is input, its type is preserved """
+        x = np.array([0, 10, 20])
+        y = x / 10
+        x_edges = np.array([0, 5, 15, 20])
+        y_edges = x_edges / 10
+        hist, edges = histogramdd((x, y), bins=(x_edges, y_edges))
+
+        assert_equal(edges[0].dtype, x_edges.dtype)
+        assert_equal(edges[1].dtype, y_edges.dtype)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: These fail on master, as `edges[i].dtype` is `float64`

File: numpy/lib/tests/test_histograms.py [hunk 3]
```diff
@@ -660,3 +658,40 @@ def test_finite_range(self):
...
+
+    def test_large_integers(self):
+        big = 2**60  # Too large to represent with a full precision float
+
+        x = np.array([0], np.int64)
+        x_edges = np.array([-1, +1], np.int64)
+        y = big + x
+        y_edges = big + x_edges
+
+        hist, edges = histogramdd((x, y), bins=(x_edges, y_edges))
```
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: This also fails on master with
```
ValueError: `bins[0]` must be strictly increasing, when an array
```

File: numpy/lib/tests/test_histograms.py [hunk 4]
```diff
@@ -660,3 +658,40 @@ def test_finite_range(self):
...
         assert_raises(ValueError, histogramdd, vals,
                       range=[[0.0, 1.0], [np.nan, 0.75], [0.25, 0.5]])
+
+    def test_equal_edges(self):
+        """ Test that adjacent entries in an edge array can be equal """
+        x = np.array([0, 1, 2])
+        y = np.array([0, 1, 2])
+        x_edges = np.array([0, 2, 2])
+        y_edges = 1
+        hist, edges = histogramdd((x, y), bins=(x_edges, y_edges))
```
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: On master, this test crashes with
```
ValueError: `bins[0]` must be strictly increasing, when an array
```

File: numpy/lib/tests/test_histograms.py [hunk 5]
```diff
@@ -646,7 +644,7 @@ def test_rightmost_binedge(self):
         bins = [[0., 0.5, 1.0]]
         hist, _ = histogramdd(x, bins=bins)
         assert_(hist[0] == 0.0)
-        assert_(hist[1] == 1.)
+        assert_(hist[1] == 0.0)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: This test was weird. `1.0000000001 <= 1.0` should be false.

**Comment by charris** [comment]
Hmm, this does change behavior and might cause problems downstream, so it probably needs a compatibility note in the release notes.

**Comment by eric-wieser** [comment]
I'll add one. I don't think the change is substantially different from the rounding fixes added to `np.histogram` as part of 1.15, so I'll add an entry for both

**Comment by eric-wieser** [comment]
@charris: Added, along with `np.histogram` notes. If this doesn't make 1.15, then the release notes still need to (with modifications).

**Pull request merged**